File: MARCH/06-03-2026/Smallest window containing all characters.py
from collections import Counter
class Solution:
    def minWindow(self, s, p):
        need = Counter(p)
        window = {}
        
        have = 0
        need_count = len(need)
        
        left = 0
        min_len = float("inf")
        result = ""
        
        for right in range(len(s)):
            char = s[right]
            window[char] = window.get(char, 0) + 1
        
            if char in need and window[char] == need[char]:
                have += 1
        
            # shrink window
            while have == need_count:
                if (right - left + 1) < min_len:
                    min_len = right - left + 1
                    result = s[left:right+1]
        
                window[s[left]] -= 1
                if s[left] in need and window[s[left]] < need[s[left]]:
                    have -= 1
        
                left += 1
        
        return result

File: MARCH/06-03-2026/Smallest window containing all characters.cpp
class Solution {
  public:
    string minWindow(string &s, string &p) {
        int m=s.length(), n=p.length();
        if(m<n) return "";
        
        int minLen=1e9, left=0, start=-1, cnt=0;
        unordered_map<char,int> mp;
        
        for(int i=0;i<n;i++) mp[p[i]]++;
        
        for(int right=0;right<m;right++){
            if(mp[s[right]]>0) cnt++;
            mp[s[right]]--;
            if(cnt==n){
                while(mp[s[left]]<0) {
                    mp[s[left]]++;
                    left++;
                }
                if(minLen>right-left+1){
                    minLen=right-left+1;
                    start=left;
                }
            }
        }
        return minLen==1e9?"":s.substr(start,minLen);
    }
};